src/main/java/ch/njol/skript/sections/SecFilter.java

```diff
@@ -0,0 +1,191 @@
+import ch.njol.skript.lang.Condition;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.InputSource;
+import ch.njol.skript.lang.Section;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.TriggerItem;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.skript.variables.Variables;
+import ch.njol.util.Kleenean;
+import ch.njol.util.Pair;
+import ch.njol.util.StringUtils;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.StreamSupport;
+
+@Name("Filter (Section)")
+@Description({
+	"Filters a variable list based on the supplied conditions. Unlike the filter expression, this effect " +
+	"maintains the indices of the filtered list.",
+	"It also supports filtering based on meeting any of the given criteria, rather than all, like multi-line if statements."
+})
+@Examples({
+	"set {_a::*} to integers between -10 and 10",
+	"",
+	"filter {_a::*} to match:",
+		"\tinput is a number",
+		"\tmod(input, 2) = 0",
+		"\tinput > 0",
+	"",
+	"send {_a::*} # sends 2, 4, 6, 8, and 10",
+})
+@Since("INSERT VERSION")
+public class SecFilter extends Section implements InputSource {
+
+	static {
+		Skript.registerSection(SecFilter.class,
+				"filter %objects% to match [:any|all]");
+		if (!ParserInstance.isRegistered(InputSource.InputData.class))
+			ParserInstance.registerData(InputSource.InputData.class, InputSource.InputData::new);
+	}
+
+	@UnknownNullability
+	private Variable<?> unfilteredObjects;
+	private final List<Condition> conditions = new ArrayList<>();
+	private boolean isAny;
+
+	@Nullable
+	private Object currentValue;
+	@UnknownNullability
+	private String currentIndex;
+	private final Set<ExprInput<?>> dependentInputs = new HashSet<>();
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, SectionNode sectionNode, List<TriggerItem> triggerItems) {
+		if (expressions[0].isSingle() || !(expressions[0] instanceof Variable)) {
+			Skript.error("You can only filter list variables!");
+			return false;
+		}
+		unfilteredObjects = (Variable<?>) expressions[0];
+		isAny = parseResult.hasTag("any");
+
+		// Code pulled from SecConditional
+		ParserInstance parser = getParser();
+		if (sectionNode.isEmpty()) {
+			Skript.error("filter sections must contain at least one condition");
+			return false;
+		}
+		InputSource.InputData inputData = getParser().getData(InputSource.InputData.class);
+		InputSource originalSource = inputData.getSource();
+		inputData.setSource(this);
+		try {
+			for (Node childNode : sectionNode) {
+				if (childNode instanceof SectionNode) {
```

Comment:
do you think it would be better to check that it's *not* a simple node?

Reply:
yes good idea